pkg/metrics/sketch_series.go

```diff
@@ -78,6 +100,128 @@ func (sl SketchSeriesList) MarshalJSON() ([]byte, error) {
 }
 
+// MarshalSplitCompress uses the stream compressor to marshal and compress sketch series payloads.
+// If a compressed payload is larger than the max, a new payload will be generated. This method returns a slice of
+// compressed protobuf marshaled gogen.SketchPayload objects. gogen.SketchPayload is not directly marshaled - instead
+// it's contents are marshaled individually, packed with the appropriate protobuf metadata, and compressed in stream.
+// The resulting payloads (when decompressed) are binary equal to the result of marshaling the whole object at once.
+func (sl SketchSeriesList) MarshalSplitCompress(bufferContext *marshaler.BufferContext) ([]*[]byte, error) {
+	// The Metadata field of gogen.SketchPayload is never written to - so pack an empty metadata as the footer
+	footer := []byte{0x12, 0}
+
+	bufferContext.CompressorInput.Reset()
+	bufferContext.CompressorOutput.Reset()
+
+	compressor, e := stream.NewCompressor(bufferContext.CompressorInput, bufferContext.CompressorOutput, []byte{}, footer, []byte{})
+	if e != nil {
+		return nil, e
+	}
+	payloads := []*[]byte{}
+
+	dsl := make([]gogen.SketchPayload_Sketch_Dogsketch, 1)
+	for _, ss := range sl {
+		if len(ss.Points) > cap(dsl) {
+			dsl = make([]gogen.SketchPayload_Sketch_Dogsketch, len(ss.Points))
+		}
+
+		for i, p := range ss.Points {
+			b := p.Sketch.Basic
+			k, n := p.Sketch.Cols()
+			dsl[i] = gogen.SketchPayload_Sketch_Dogsketch{
+				Ts:  p.Ts,
+				Cnt: b.Cnt,
+				Min: b.Min,
+				Max: b.Max,
+				Avg: b.Avg,
+				Sum: b.Sum,
+				K:   k,
+				N:   n,
+			}
+		}
+
+		sketch := gogen.SketchPayload_Sketch{
+			Metric:      ss.Name,
+			Host:        ss.Host,
+			Tags:        ss.Tags,
+			Dogsketches: dsl[:len(ss.Points)],
+		}
+
+		// Pack the protobuf metadata - see SketchPayload.MarshalTo in agent_payload.pb.go for reference.
+		metadataSize := 0
+		// Magic number that occurs before the varint encoding
+		bufferContext.PrecompressionBuf[metadataSize] = 0xa
+		metadataSize++
+		metadataSize = encodeVarintAgentPayload(bufferContext.PrecompressionBuf, metadataSize, uint64(sketch.Size()))
+
+		// Resize the pre-compression buffer if needed
+		totalItemSize := sketch.Size() + metadataSize
+		if totalItemSize > cap(bufferContext.PrecompressionBuf) {
+			bufferContext.PrecompressionBuf = append(bufferContext.PrecompressionBuf, make([]byte, totalItemSize-cap(bufferContext.PrecompressionBuf))...)
+			bufferContext.PrecompressionBuf = bufferContext.PrecompressionBuf[:cap(bufferContext.PrecompressionBuf)]
+		}
+
+		// Marshal the sketch to the precompression buffer after the metadata
+		_, e := sketch.MarshalTo(bufferContext.PrecompressionBuf[metadataSize:])
+		if e != nil {
+			return nil, e
+		}
+
+		// Compress the protobuf metadata and the marshaled sketch
+		e = compressor.AddItem(bufferContext.PrecompressionBuf[:totalItemSize])
+		if e == stream.ErrPayloadFull {
+			tlmPayloadFull.Inc()
+		}
+		if e == stream.ErrItemTooBig || e == stream.ErrPayloadFull {
```

Comment:
Unless I'm missing something, I think you misunderstood what `stream.ErrItemTooBig` means: I think it means that the item alone couldn't even fit in a new payload of its own, because the compressor expects it to be larger than the max allowed payload size (compressed or uncompressed).

So if you got `stream.ErrItemTooBig` on the first try for a given item, you'll always get this same error on the second try. So you should just drop the item after the first try (and track that with telemetry)

Reply:
Fixed and added a test for this case.